url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/spatialindex/rtree/RTree.java

```diff
@@ -0,0 +1,1674 @@
+
+		boolean bRet = false;
+
+		Stack<Integer> pathBuffer = new Stack<Integer>();
+
+		Node root = readNode(m_rootID);
+		Leaf l = root.findLeaf(mbr, id, pathBuffer);
+
+		if (l != null) {
+			l.deleteData(id, pathBuffer, changedNodeID);
+			m_stats.m_data--;
+			bRet = true;
+		}
+
+		return bRet;
+	}
+
+	/**
+	 * write the node into file
+	 * 
+	 * @param n
+	 * @return
+	 * @throws IllegalStateException
+	 */
+	protected int writeNode(Node n) throws IllegalStateException {
+		byte[] buffer = null;
+
+		try {
+			//store node into byte
+			buffer = n.store();
+		} catch (IOException e) {
+			System.err.println(e);
+			throw new IllegalStateException("writeNode failed with IOException");
+		}
+
+		int page;//select which page should stored
+		if (n.m_identifier < 0) {
+			page = IStorageManager.NewPage;
+		} else {
+			page = n.m_identifier;
+		}
+
+		try {
+			//store byte into storage manager
+			page = m_pStorageManager.storeByteArray(page, buffer);
+		} catch (InvalidPageException e) {
+			System.err.println(e);
+			throw new IllegalStateException("writeNode failed with InvalidPageException");
+		}
+
+		if (n.m_identifier < 0) {
+			//change id from -1 to reserved page
+			n.m_identifier = page;
+			//change data in STAT
+			m_stats.m_nodes++;
+			int i = m_stats.m_nodesInLevel.get(n.m_level).intValue();
+			m_stats.m_nodesInLevel.set(n.m_level, new Integer(i + 1));
+		}
+
+		m_stats.m_writes++;
+
+		// TODO What this means? Any class inherit from INodeCommand?
+		for (int cIndex = 0; cIndex < m_writeNodeCommands.size(); cIndex++) {
+			m_writeNodeCommands.get(cIndex).execute(n);
+		}
+
+		return page;
+	}
+
+	public synchronized Node readNode(int id) {
+		byte[] buffer;
+		DataInputStream ds = null;
+		int nodeType = -1;
+		Node n = null;
+
+		try {
+			buffer = m_pStorageManager.loadByteArray(id);
+			ds = new DataInputStream(new ByteArrayInputStream(buffer));
+			nodeType = ds.readInt();
+
+			if (nodeType == SpatialIndex.PersistentIndex) {
+				n = new Index(this, -1, 0);
+			} else if (nodeType == SpatialIndex.PersistentLeaf) {
+				n = new Leaf(this, -1);
+			} else {
+				throw new IllegalStateException("readNode failed reading the correct node type information");
+			}
+
+			n.m_pTree = this;
+			n.m_identifier = id;
+			n.load(buffer);
+
+			m_stats.m_reads++;
+		} catch (InvalidPageException e) {
+			System.err.println(e);
+			System.err.println("id : " + id + "is not existed!");
+			return n;
+			//throw new IllegalStateException("readNode failed with InvalidPageException");
+		} catch (IOException e) {
+			System.err.println(e);
+			throw new IllegalStateException("readNode failed with IOException");
+		}
+
+		for (int cIndex = 0; cIndex < m_readNodeCommands.size(); cIndex++) {
+			m_readNodeCommands.get(cIndex).execute(n);
+		}
+
+		return n;
+	}
+
+	protected void deleteNode(Node n) {
+		try {
+			m_pStorageManager.deleteByteArray(n.m_identifier);
+		} catch (InvalidPageException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)